File: 006.ZigZag Conversion/wyrj/ZigZagConversion.js
/* 220ms */
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if (1 === numRows) {
        return s;
    }
    var ret = '';
    for (var i = 0; i < numRows; i++) {
        var j = i;
        var next = [2*(numRows-i) - 3, 2*i - 1];
        if (next[0] < 0) {
            next[0] = next[1];
        } else if (next[1] < 0) {
            next[1] = next[0];
        }
        var flag = 0;
        while (j < s.length) {
            ret += s[j];
            j += next[flag % 2] + 1;
            flag++;
        }
    }
    return ret;
};

File: 006.ZigZag Conversion/wyrj/ZigZagConversion.c
/* 8ms */
char* convert(char* s, int numRows) {
    if (1 == numRows) {
        return s;
    }
    int len = strlen(s);
    char* ret = (char *)malloc(sizeof(char)*(len+1));
    ret[len] = '\0';
    int retIndex = 0, pingpongMax = (numRows - 2) * 2;
    int i, j, next;
    for (i = 0; i < numRows; i++) {
        j = i;
        next = 2 * i - 1;
        while(j < len) {
            ret[retIndex++] = s[j];
            if (next < pingpongMax) {
                next = pingpongMax - next;
            }
            j += next + 1;
        }
    }
    return ret;
}